### PR TITLE
feat: DevOps agent built with packer

### PR DESCRIPTION
**Pull request by mamari90**
### List of changes

- updated azdoa agent to accept custom images
- created new module to build, using packer, custom images (executed manually)

### Motivation and context

the devops agents keep failing, missing some required installed programs. So we decided to create a custom image containing all the required programs, to avoid installing them every time an agent is created

### Type of changes

- [x] Add new module
- [x] Update existing module
- [ ] Remove existing module

### Does this introduce a breaking change?

- [x] Yes
- [ ] No

### Run checks

Useful commands to run checks on local machine

```sh
bash .utils/terraform_run_all.sh init local
pre-commit run -a
```

### CODE REVIEW

File: azure_devops_agent/README.md [hunk 1]
```diff
@@ -1,10 +1,18 @@
...
 
-This module allow to create an azure devops agent (vm scale set), and inside the folder scripts has a lot of utilities
 
 ## How to use
 
-```ts
+
+by default, this module creates a ScaleSet using Ubuntu22.04, without using os disk encryption, and providing SSH access using a generated ssh key
+
+```hcl
```
Comment (diegolagospagopa) [inline azure_devops_agent/README.md:15]: hcl don't exist,please use ts

File: azure_devops_agent/main.tf [hunk 2]
```diff
@@ -1,115 +1,92 @@
...
+  single_placement_group      = false
+  upgrade_mode                = "Manual"
+
+
+  extension {
+    name                 = "install_requirements"
+    publisher            = "Microsoft.Azure.Extensions"
+    type                 = "CustomScript"
+    type_handler_version = "2.0"
+    settings             = file("${path.module}/script-config.json")
```
Comment (diegolagospagopa) [inline azure_devops_agent/main.tf:81]: this extension is still needed?

File: azure_devops_agent/main.tf [hunk 3]
```diff
@@ -1,115 +1,92 @@
...
+  admin_username      = "adminuser"
+  admin_password      = var.admin_password
+
+
+  # only one of source_image_id and source_image_reference is allowed
+  source_image_id = var.source_image_name != null ? local.source_image_id : null
+
+  dynamic "source_image_reference" {
+    # only one of source_image_id and source_image_reference is allowed
+    for_each = var.image_reference != null ? [1] : []
```
Comment (diegolagospagopa) [inline azure_devops_agent/main.tf:36]: can go in conflict?

File: azure_devops_agent_custom_image/README.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+Then, you can simply pick up the built image name from the log, and configure it to be used as base image for your vm or scale set
+The image name will be found in the logs, in the following line
+
+```
+module.azdoa_custom_image.null_resource.build_packer_image (local-exec): ManagedImageName: my_image_name-v3
+```
+
+Example:
+```hcl
```
Comment (diegolagospagopa) [inline azure_devops_agent_custom_image/README.md:30]:
```suggestion
```hcl
```ts

File: azure_devops_agent_custom_image/devops_agent.json [hunk 5]
```diff
@@ -0,0 +1,6 @@
+{
+  "isPipelinesAgent": true,
+  "agentFolder": "/agent",
+  "agentDownloadUrl": "https://vstsagentpackage.azureedge.net/agent/2.202.1/vsts-agent-linux-x64-2.202.1.tar.gz",
+  "enableScriptDownloadUrl": "https://vstsagenttools.blob.core.windows.net/tools/ElasticPools/Linux/11/enableagent.sh"
+}
```
Comment (diegolagospagopa) [inline azure_devops_agent_custom_image/devops_agent.json:6]: Is this file still valid?

File: azure_devops_agent_custom_image/outputs.tf [hunk 6]
```diff
@@ -0,0 +1,10 @@
+output "custom_image_name" {
+  value       = local.target_image_name
+  description = "Name of the created image"
+}
+
+output "custom_image_id" {
+  value       = local.target_image_id
+  description = "Id of the created image"
```
Comment (diegolagospagopa) [inline azure_devops_agent_custom_image/outputs.tf:8]:
```suggestion
  description = "Azure id of the custom image you just created"
```

**Review by diegolagospagopa — CHANGES_REQUESTED** [review]
great job

File: azure_devops_agent/README.md (hunk 1 above)
Comment (mamari90) [reply azure_devops_agent/README.md:15]: it is handled by github

File: azure_devops_agent/main.tf (hunk 2 above)
Comment (mamari90) [reply azure_devops_agent/main.tf]: nope, removed

File: azure_devops_agent/main.tf (hunk 3 above)
Comment (mamari90) [reply azure_devops_agent/main.tf]: added a third variable to handle mutual exclusion

File: azure_devops_agent_custom_image/README.md (hunk 4 above)
Comment (mamari90) [reply azure_devops_agent_custom_image/README.md:30]: hcl is working fine

File: azure_devops_agent_custom_image/devops_agent.json (hunk 5 above)
Comment (mamari90) [reply azure_devops_agent_custom_image/devops_agent.json]: removed

**Pull request merged**